### DISCUSSION TITLE
Possible Issue:: Char Array converts to string possibly unusually. Demo code included.

### DISCUSSION THREAD

**Discussion by akshays2112** [dotnet/roslyn#58419] [Q&A]
Environment Basic Information to reproduce the possible issue::
C#10 (dotnet --version => 6.0.101)
Visual Studio 2022 Community Edition 64-bit (Version:: 17.0.4)

Run this code in Console Project

        int start = 2;
        Index idxStart = 2;
        int end = 5;
        Index idxEnd = 5;
        char[] array = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i' };
        Console.WriteLine(array[start..end]);
        Console.WriteLine($"{array[2..5]} = {array[start..end]} = {array[idxStart..idxEnd]}");
        Console.WriteLine($"{new string(array[start..end])} = {new string(array[idxStart..idxEnd])}");

Output::
cde
System.Char[] = System.Char[] = System.Char[]
cde = cde

Question::
The problem when using string interpolation $ instead of returning "cde" for the char[] it returns the type name "System.Char[]".
This could cause problems in lots of places most obvious of which is logging for most programmers who would not expect this behavior.
Would it be worth it to include this in the Analyzer to check for this problem and report it to the developer as a Warning for them to decide if this requires fixing or this is intentionally what they want?

Thanks in advance.

**Comment by PathogenDavid** [comment]
This is happening because `Console.WriteLine` has [an overload](https://docs.microsoft.com/en-us/dotnet/api/system.console.writeline?view=net-6.0#System_Console_WriteLine_System_Char___) specifically for printing the contents of character arrays.

In contrast using them with interpolated strings is calling `ToString` on the array which simply prints `System.Char[]` since that's what `ToString` does for all arrays.

> Would it be worth it to include this in the Analyzer to check for this problem and report it to the developer as a Warning for them to decide if this requires fixing or this is intentionally what they want?

Personally I don't think it's justified for this specific case, but you can [propose an rule be created in the roslyn-analyzers repo](https://github.com/dotnet/roslyn-analyzers/issues/new/choose).

Perhaps a more general purpose rule might be one which identifies the use of `ToString` on types which do not override `ToString`.
